internal/orchestrator/ray_orchestrator.go

```diff
@@ -452,6 +461,255 @@ func (o *RayOrchestrator) getDashboardService(ctx context.Context) (dashboard.Da
 	return dashboardService, nil
 }
 
+// CreateEndpoint deploys a new endpoint using Ray Serve.
+func (o *RayOrchestrator) CreateEndpoint(endpoint *v1.Endpoint) (*v1.EndpointStatus, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), o.opTimeout.CommonTimeout)
+	defer cancel()
+
+	// pre-check related resources
+	cluster, err := o.storage.ListCluster(storage.ListOption{
+		Filters: []storage.Filter{
+			{
+				Column:   "metadata->name",
+				Operator: "eq",
+				Value:    endpoint.Spec.Cluster,
+			},
+		},
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list cluster")
+	}
+
+	if len(cluster) == 0 {
+		return nil, errors.New("cluster " + endpoint.Spec.Cluster + " not found")
+	}
+
+	engine, err := o.storage.ListEngine(storage.ListOption{
+		Filters: []storage.Filter{
+			{
+				Column:   "metadata->name",
+				Operator: "eq",
+				Value:    endpoint.Spec.Engine.Engine,
+			},
+		},
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list engine")
+	}
+
+	if len(engine) == 0 {
+		return nil, errors.New("engine " + endpoint.Spec.Engine.Engine + " not found")
+	}
+
+	if engine[0].Status.Phase != v1.EnginePhaseCreated {
+		return nil, errors.New("engine " + endpoint.Spec.Engine.Engine + " not ready")
+	}
+
+	versionMatched := false
+
+	for _, v := range engine[0].Spec.Versions {
+		if v.Version == endpoint.Spec.Engine.Version {
+			versionMatched = true
+			break
+		}
+	}
+
+	if !versionMatched {
+		return nil, errors.New("engine " + endpoint.Spec.Engine.Engine + " version " + endpoint.Spec.Engine.Version + " not found")
+	}
+
+	modelRegistry, err := o.storage.ListModelRegistry(storage.ListOption{
+		Filters: []storage.Filter{
+			{
+				Column:   "metadata->name",
+				Operator: "eq",
+				Value:    endpoint.Spec.Model.Registry,
+			},
+		},
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list model registry")
+	}
+
+	if len(modelRegistry) == 0 {
+		return nil, errors.New("model registry " + endpoint.Spec.Model.Registry + " not found")
+	}
+
+	if modelRegistry[0].Status.Phase != v1.ModelRegistryPhaseCONNECTED {
+		return nil, errors.New("model registry " + endpoint.Spec.Model.Registry + " not ready")
+	}
+
+	// call ray dashboard API
+
+	dashboardService, err := o.getDashboardService(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get dashboard service for creating endpoint")
+	}
+
+	currentAppsResp, err := dashboardService.GetServeApplications()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get current serve applications")
+	}
+
+	newApp := dashboard.EndpointToApplication(endpoint, &modelRegistry[0])
+
+	// Build the list of applications for the PUT request
+	updatedAppsList := make([]dashboard.RayServeApplication, 0, len(currentAppsResp.Applications)+1)
+	for _, appStatus := range currentAppsResp.Applications {
+		updatedAppsList = append(updatedAppsList, *appStatus.DeployedAppConfig)
+	}
+
+	updatedAppsList = append(updatedAppsList, newApp)
```

Comment:
need to determine whether the app currently exists?

Reply:
yep, updated in the new commit